Create CPS file with Python Scripts

andersonfrailey opened the pull request:
I want to start by emphasizing that this is still very far from completion. Looking at the weighted totals it's obvious that there is still work to be done. As I mentioned in the Tax-Calculator maintainers call last night, I wanted to open this PR so that others can see what I'm doing and offer feedback/suggestions as I go. With all that being said, here's a bit on what's going on.

To start, I download the CPS files from [NBER](https://www.nber.org/data/current-population-survey-data.html). I wrote a few scripts to handle reading the .DAT files and converting them to .CSV files: `transform_sas.py`, `template.txt`, `cpsmar*.py`. I'll be happy to explain what they each do if there's interest. I'm using the March 2013, 2014, and 2015 CPS supplements to be consistent with our current data.

After that, I take each of the CPS files and form tax units. I only look within households for tax unit formation.

Starting with the reference person for each family, I loop through everyone in the household. If they are not already part of a tax unit (as a spouse, dependent, or head of the unit) I form a new tax unit with them. I then search through everyone in the household to see if they have a spouse or dependents to claim. I also check all dependents to see if they are required to file. After all this is done, the tax units are exported.

I take a DataFrame with all of the units and run it through a final prep step to clean up the file, and the result is a CSV that should be usable with taxcalc (haven't tried it yet).

I've written a short script `validate.py` that does some very basic testing of the file and the results are promising. There are two things it's checking for: 1) has everyone in the household been accounted for and 2) is income in the household being counted correctly. 1) is checked by making sure the number of people appearing in the `nu18`, `n1820`, and `n21` variables are equal to the actual number of people in the household within those age ranges. 2) is checked by looking at a number of income variables <sup>1</sup> and asserting that the total in every tax unit in that household is less than or equal to the total amount in the household. I allow it to be less than the total in the house because there could be some dependents who have a little income, but not enough to have to file. All of these checks pass with no errors.

In terms of weighted totals, there are mixed results. Some variables, like wages, are within 2% of the PUF. Others, like farm income, are significantly off. Given that the basic validation testing I've done has passed, It's possible that these issues could be solved by reweighing. I will try and figure out the best way for sharing the current weighted totals and comparisons to the PUF and CPS files.

Next steps are to try and replicate the imputations currently being done for the CPS so that we have all the same variables. Add the benefit data from C-TAM. And do general code improvements.

<sup>1</sup> Wages, interest income, self-employment income, farm income, dividend income, unemployment compensation, pensions, and alimony

cc @MattHJensen @MaxGhenis @martinholmer

MaxGhenis commented:
>2) is checked by looking at a number of income variables 1 and asserting that the total in every tax unit in that household is less than or equal to the total amount in the household. I allow it to be less than the total in the house because there could be some dependents who have a little income, but not enough to have to file.

Do the scripts create tax units for nonfilers, e.g. with a `filer` flag set to `False`?

andersonfrailey commented:
@MaxGhenis at this point they do not. My reservation with adding a `filer` flag is that as filing requirements change it can be misleading. I think a `filer` variable would be more useful if it was determined in tax-calc where it could be determined based on the year and policy the model is being run for.

MaxGhenis commented:
OK why discard dependents who don't have enough to file? Should that instead be done in taxcalc (if at all)? I thought one of the valuable aspects of the CPS file is capturing all income.

andersonfrailey commented:
Are you proposing making all dependents into their own tax unit, or just those with income?

MaxGhenis commented:
How does the SAS-produced CPS file do it?

martinholmer commented:
@andersonfrailey,  What you're doing in taxdata #314 is a good idea.  Thank you for undertaking this project.

My main questions is about this statement:
>  I'm using the March 2013, 2014, and 2015 CPS supplements to be consistent with our current data.

Are you doing this only to be able to compare results from the current `cps.csv.gz` file or do you see this as the end product of your work?  Is makes sense to me to check the new Python approach with the old SAS approach by having the two approaches produce the same data file.  But once you have confidence in the logic of your Python approach, wouldn't it make sense to switch away from the Mar13+Mar14+Mar15 data?  Wouldn't more recent data be better?  And why exactly do we need three Mar files?   Why not just one (the most recent CPS available)?  I never understood why John made such a large file; it doesn't add anything to Tax-Calculator as far as I can see, and it does impose costs in terms of larger memory requirements and slower execution.

andersonfrailey commented:
@MaxGhenis 
>How does the SAS-produced CPS file do it?

The SAS-produced file only includes dependents as tax units if they are required to file themselves. All others are counted up and represented by the tax unit claiming them, but any potential income they may have is not included anywhere.

@martinholmer 
>Are you doing this only to be able to compare results from the current cps.csv.gz file or do you see this as the end product of your work?

I'm doing this just to compare my results to the current `cps.csv.gz` file.

> Is makes sense to me to check the new Python approach with the old SAS approach by having the two approaches produce the same data file. But once you have confidence in the logic of your Python approach, wouldn't it make sense to switch away from the Mar13+Mar14+Mar15 data? Wouldn't more recent data be better?

Yes. I believe that it would be better to use a more recent file once I have the Python approach ironed out. The CPS we end up picking should be a discussion between the TaxData, Tax-Calculator, and C-TAM projects (cc @Amy-Xu ).

>And why exactly do we need three Mar files? Why not just one (the most recent CPS available)? I never understood why John made such a large file; it doesn't add anything to Tax-Calculator as far as I can see, and it does impose costs in terms of larger memory requirements and slower execution.

My understanding is that three files were used in an effort to increase the sample size. I don't believe that it is necessary to continue using three separate files. I'll compare what we get when using one file versus three so we can get a firm view of what, if any, advantage there is to using three files.

MaxGhenis commented:
>The SAS-produced file only includes dependents as tax units if they are required to file themselves. All others are counted up and represented by the tax unit claiming them, but any potential income they may have is not included anywhere.

OK thanks. It'd be useful to capture all income at some point to evaluate reforms that affect filing rules, but makes sense to keep that out of this PR.

>My understanding is that three files were used in an effort to increase the sample size.

FWIW there's precedent for this in the [Supplemental Poverty Measure](https://www.census.gov/content/dam/Census/library/publications/2018/demo/p60-265.pdf), where state-level poverty estimates are only published using 3-year estimates, though most of the SPM and other similar analyses I've seen just use one year. Whether this is analytically better for taxcalc seems like an empirical question, but IMO a one-year file would be easiest, also because it's simpler to merge with the original CPS.
![image](https://user-images.githubusercontent.com/6076111/58580565-4030db80-8202-11e9-8eb0-3875ceeb4b99.png)

andersonfrailey commented:
My latest commits build on the work that's been done up to this point. In them I've added the C-TAM benefits, enhanced the validation tests, and added state level income targeting*. At this point I'm still working with the 2013, 2014, and 2015 CPS files so that we can compare to the current version of `cps.csv.gz`.

When comparing totals between the Python generated CPS, the PUF, and the current CPS file, there are still some odd results. You can download a comma-separated text file to see this
[here](https://github.com/PSLmodels/taxdata/files/3339835/pycpstotals.txt). EIC, elderly_dependents, and the benefit variables particularly stand out.

It would be helpful for me to have the logic I'm using to determine who is a dependent and who is EIC eligible reviewed. The code I wrote for both could definitely be improved.

The next step is to add the imputations that are on `cps.csv.gz`, but not in the Python generated CPS yet.

Despite some odd aggregate numbers, I'm fairly confident in what I have so far. In the validation tests, I make sure that everyone in the household is accounted for in one of the age variables (`nu18`, `n1820`, and `n21`) and that all income and benefits accrued by the household gets counted. Each of these tests pass without issue.

*For state level targeting, I take [state level IRS data](https://www.irs.gov/statistics/soi-tax-stats-historic-table-2) on the amount of certain income variables and adjust incomes in the CPS so that these totals are hit for each state. We do this when making the current version of `cps.csv.gz`. I haven't yet restricted this targeting to be done using just those who would file taxes.

cc @martinholmer @MaxGhenis @MattHJensen

MaxGhenis commented on cps_data/pycps/finalprep.py:
```diff
@@ -0,0 +1,116 @@
+    )
+
+    # rename variables
+    renames = {
```
```suggestion
    RENAMES = {
```

MaxGhenis commented on cps_data/pycps/pycps.py:
```diff
@@ -0,0 +1,277 @@
+    https://www.irs.gov/credits-deductions/individuals/earned-income-tax-credit/qualifying-child-rules
+    """
+    # relationship test
+    relationship = ((person["a_exprrp"] == 5) |
```
```suggestion
    relationship = person.a_exprrp.isin([5, 7, 9, 11])
```

MaxGhenis commented on cps_data/pycps/pycps.py:
```diff
@@ -0,0 +1,277 @@
+    if person["a_ftpt"] == 1:
+        eic_max_age = 24
+    # person is between 1 and the max age
+    age = 0 <= person["a_age"] <= eic_max_age
```
```suggestion
    age = (0 <= person["a_age"] <= eic_max_age)
```

MaxGhenis replied:
(clarity only)

MaxGhenis commented on cps_data/pycps/pycps.py:
```diff
@@ -0,0 +1,277 @@
+    cps["d_flag"] = False  # dependent flag
+    cps["hhid"] = cps["h_seq"]
+    # calculate earned and unearned income
+    earned_inc_vars = [
```
```suggestion
    EARNED_INC_VARS = [
```

MaxGhenis replied:
also below

MaxGhenis commented on cps_data/pycps/targeting.py:
```diff
@@ -0,0 +1,66 @@
+    state_data = state_data[state_data["AGI_STUB"] == 0].copy()
+
+    # map fips codes and to state abbreviations
+    fips_dict = {'AK': 2, 'AL': 1, 'AR': 5, 'AZ': 4, 'CA': 6, 'CO': 8, 'CT': 9,
```
```suggestion
    FIPS_DICT = {'AK': 2, 'AL': 1, 'AR': 5, 'AZ': 4, 'CA': 6, 'CO': 8, 'CT': 9,
```

MaxGhenis commented on cps_data/pycps/targeting.py:
```diff
@@ -0,0 +1,66 @@
+                 'WV': 54, 'WY': 56}
+    # map income variables in the CPS and IRS data
+    # TODO: Add imputed variables
+    var_map = {
```
```suggestion
    VAR_MAP = {
```

MaxGhenis commented on cps_data/pycps/targeting.py:
```diff
@@ -0,0 +1,66 @@
+        for state, fips in fips_dict.items():
+            sub_cps = cps[cps["fips"] == fips]
+            target = state_data[var][state] * 1000  # scale up IRS data
+            cps_uw_total = sub_cps[cps_vars].sum(axis=1)
```
Add a comment noting this is a vector with the total for each tax unit (I was confused at first).

MaxGhenis commented on cps_data/pycps/taxunit.py:
```diff
@@ -0,0 +1,195 @@
+INCOME_TUPLES = [
```
Philosophical question about this file: Why make a new `TaxUnit` object, rather than a DataFrame?

andersonfrailey replied:
This was mostly to make it easier for me to conceptualize what was happening. Treating the tax unit as its own object made it easier for me to form the logic of creating a tax unit, without severely complicating transforming that `TaxUnit` object into a row in the final DataFrame.

andersonfrailey commented:
Thanks for the comments, @MaxGhenis. I'll be working through them in upcoming commits

andersonfrailey commented on cps_data/pycps/benefits.py:
```diff
@@ -0,0 +1,119 @@
+
+    tanf_path = Path(data_path, f"TANF_Imputation_{year}.csv")
+    tanf = pd.read_csv(tanf_path, usecols=["peridnum", "tanf_impute"])
+    # drop duplicated people in tanf
```
@Amy-Xu could provide a better answer to this than I could

MaxGhenis commented on cps_data/pycps/benefits.py:
```diff
@@ -0,0 +1,118 @@
+                       ["fh_seq", "ffpos", "housing_impute"])
+    tanf = read_ben("TANF_Imputation_", ["peridnum", "tanf_impute"])
+    # drop duplicated people in tanf
+    tanf = tanf.drop_duplicates("peridnum")
```
```suggestion
    tanf.drop_duplicates("peridnum", inplace=True)
```

MaxGhenis commented on cps_data/pycps/benefits.py:
```diff
@@ -0,0 +1,118 @@
+
+    data["housing_ben"] *= 12
+    # Convert benefit data to integers
+    ben_vars = [
```
```suggestion
    BEN_VARS = [
```

MaxGhenis commented on cps_data/pycps/benefits.py:
```diff
@@ -0,0 +1,118 @@
+        "mcaid_ben", "mcare_ben", "ssi_ben", "snap_ben", "vet_ben", "tanf_ben",
+        "wic_ben", "housing_ben", "e02400", "e02300", "other_ben"
+    ]
+    for var in ben_vars:
```
```suggestion
    for var in BEN_VARS:
```

MaxGhenis commented on cps_data/pycps/finalprep.py:
```diff
@@ -0,0 +1,116 @@
+        'housing_ben', "linenos"
+    ]
+
+    drop_vars = []
```
You can now delete lines 31-34

MaxGhenis commented on cps_data/pycps/pycps.py:
```diff
@@ -0,0 +1,274 @@
+    # add apply create_unit with a progress bar
+    tqdm.pandas(desc=str(year))
+    units = cps.groupby("h_seq").progress_apply(create_units, year=year - 1)
+    TAX_UNITS = []
```
I don't think this is a constant as the capitalization suggests

andersonfrailey commented:
BTW, I'm planning on using the same weighting methodology as we currently use for the CPS file. If anyone else has another method they'd like to try, I'd be happy to experiment with it.

MaxGhenis commented:
>BTW, I'm planning on using the same weighting methodology as we currently use for the CPS file. If anyone else has another method they'd like to try, I'd be happy to experiment with it.

Just throwing out https://github.com/PSLmodels/taxdata/issues/323 to keep the weights for tax units in the same household equal, so that they can be reaggregated to households.

andersonfrailey commented:
I just pushed a bunch of new code related to imputations. I've just copied our current imputation methods, but I believe they could be improved on. This is something that should be explored in the future.

andersonfrailey commented:
Latest commit just refactors the logic to create the tax units by doing as much as possible with dictionaries, instead of DataFrames. This has brought the runtime of the tax unit creation process down to around 20 seconds per year. All other logic is unchanged.

andersonfrailey commented:
I was able to use the linear programming model @yimengyin16 used in his presentation linked to in issue #326 and generate weights for the python generated CPS file. the new files through taxcalc to see how everything compares to the current PUF and CPS file. Will do later today. 

I'm mostly confident in the logic that creates the tax unit, but some of the new benefit totals seem odd to me.  I've attached output from the test failure and you can see which ones look most off. I'm most concerned with medicare, medicaid, SNAP, TANF, and housing. Later today I'm going to dig more into the C-TAM data to see if things like the maximum value for medicare are legit or a bug with the new code. 
<img width="406" alt="Screen Shot 2019-12-29 at 11 19 12 AM" src="https://user-images.githubusercontent.com/20684675/71560251-0c829200-2a2d-11ea-8e3c-938730b56dfe.png">

Overall though, I'd say things are looking up for this project. As always, open to any feedback y'all have.

andersonfrailey commented:
Here's a [notebook](https://nbviewer.jupyter.org/github/andersonfrailey/Notebook-Uploads/blob/master/comparison.ipynb) that does a short comparison of the python generated CPS file and the current CPS file and PUF. The most obvious difference is that the python generated CPS shows consistently lower income tax liabilities than the other two files.

I also looked into the high benefit numbers and they match up with C-TAM data. And the final CPS file also passes all of the validation tests I wrote for checking income and benefit totals in each tax unit.

As always, happy to hear any feedback/suggestions.

MaxGhenis commented:
Thanks @andersonfrailey, what do you think is driving the lower income tax in the new CPS? For 2019, it looks like $1.1T vs. $1.4T in the old CPS and $1.7T in the PUF. Payroll tax totals less than the old CPS but similar to the PUF so that doesn't seem like an issue.

Cell 19 shows that the new CPS tax units skew lower-income than the new ones, but I couldn't tell how much of the income tax difference that accounts for.

Could you share the grand total weight for each of the 3 sources? Cells 19-20 suggest they might differ by 10% or more.

If they do differ substantially, it could be interesting to see how a different version of the CPS compares, limiting to filers and scaling up the weights to match the PUF total weight.

andersonfrailey commented:
@MaxGhenis Here are the total numbers:

|         | Combined     | IITax        | PTax         |
|---------|--------------|--------------|--------------|
| PUF     | 39873.002638 | 23445.090884 | 16427.911754 |
| Old CPS | 37745.819397 | 20543.109070 | 17202.710328 |
| New CPS | 32697.887604 | 16171.936765 | 16525.950839 |

So the ten year aggregate income tax totals are very far off. 

What I've noticed is that the biggest difference in income tax liability between the new and old (current) CPS file is in the $200-500K expanded income bin. That gap makes up about 47% of the total gap for 2019.
![visualization](https://user-images.githubusercontent.com/20684675/72487239-616b2a00-37db-11ea-8e6f-7a486dfcdf4d.png)

But when you look at total taxable income, the new CPS file actually lines up pretty well with both the PUF and the old CPS.

![visualization-2](https://user-images.githubusercontent.com/20684675/72487365-c6bf1b00-37db-11ea-9773-0423ebd69ab5.png)

I realized last night that there's an issue with how I did state level targeting. There's one part of the process where we use state level IRS data to scale up or down people's income to meet the totals reported by the IRS in each state. This is something we do with the current CPS as well. But, rather than only using the people who we've determined to be filers to determine these scaling factors, I use everyone in the CPS. It's possible that this is causing the scaling factors to be too small.

I have no idea if this will actually solve the problem, but I'll re-run everything after the switch has been made and we'll see what happens.

andersonfrailey commented:
Update: It did not change total taxes very much so there's still work to do. But I am excited that all the files needed to create the CPS file and weights run in ~15 minutes rather than 6+ hours!

MaxGhenis commented:
Could you share the total number of tax units in the 3 files?

andersonfrailey commented:
<img width="447" alt="Screen Shot 2020-03-16 at 10 29 14 PM" src="https://user-images.githubusercontent.com/20684675/76816043-93960980-67d5-11ea-9d6c-dad478267cf6.png">
Here's info on the number of tax units, @MaxGhenis.

There are a ton more filers in the python generated version of the CPS, and I'm still trying to figure out what exactly is causing that. I think I'm much more strict about how tax units are formed than the SAS code has been. For example, there's a  household in the 2015 CPS containing the reference person, their two kids, and an unmarried partner. In the SAS produced CPS, this is one tax unit with the unmarried partner filing as the head of household because they're the only one with income. In the Python CPS, this is two units, both single filers: the first is the reference person and their children (there's no income in this unit), and the unmarried partner. I've been doing some research on the exact rules for forming a tax unit and how I need to change my code accordingly.

In the meantime, I fixed an issue that marked married people whose spouse absent from the file as single filers, rather than joint (according to the documentation, being absent from the file is distinct from being separated). I've added the income distribution adjustments that we've been doing, and added state taxes after credits to the file as `e18400`.

These fixes closed some of the gap between combined tax liabilities, but we're still not looking great:

| | 2019             | 2020    | 2021    | 2022    | 2023    | 2024    | 2025    | 2026    | 2027    | 2028    | 2029    |
|------------------|---------|---------|---------|---------|---------|---------|---------|---------|---------|---------|---------|
| puf_combined     | 2863.51 | 2988.47 | 3105.39 | 3221.69 | 3351.85 | 3500.75 | 3656.24 | 4018.88 | 4198.20 | 4387.00 | 4581.02 |
| old_cps_combined | 2615.40 | 2774.24 | 2917.89 | 3052.38 | 3191.31 | 3338.06 | 3482.61 | 3832.32 | 4001.00 | 4178.88 | 4361.73 |
| new_cps_combined | 2433.51 | 2550.49 | 2660.41 | 2778.79 | 2895.14 | 3019.66 | 3152.28 | 3436.78 | 3585.61 | 3741.84 | 3904.89 |

andersonfrailey commented:
I think I've got it. Played with the imputations and the state targeting a bunch and now the 10 year total tax liability is very close to the current CPS. Total income tax liability is about 0.5% less than the current CPS and payroll tax liability is about 4% less than the current CPS, but right on payroll tax liability in the PUF. Here's a plot of them all
![combined](https://user-images.githubusercontent.com/20684675/84560405-50cacb80-ad09-11ea-8cb1-e5a34533f853.png)

Not sure why the build is failing in Travis, but all the tests pass locally.

cc @MattHJensen @MaxGhenis @Peter-Metz

MaxGhenis commented:
Awesome! The Travis [build](https://travis-ci.com/github/PSLmodels/taxdata/builds/171191092) is failing on the environment:
```
$ conda env update -f environment.yml
SpecNotFound: Can't process without a name
The command "conda env update -f environment.yml" failed and exited with 1 during .
```
I think this is because the `pip` portion needs to have the dash and indents ([docs](https://conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html#create-env-file-manually)); a correct specification (also avoiding the warning requesting `pip` to be installed) would replace:

https://github.com/PSLmodels/taxdata/blob/afd2621ae6b930258f08056967515f1cb4d00e68/environment.yml#L20-L22

with
```
- pip                                                                                                                                                                                                                                        
- pip:                                                                                                                                                                                                                                       
  - rpy2                                                                                                                                                                                                                                     
  - tzlocal 
```
However, this throws this error on my machine, and might do so on Travis too:
```
Pip subprocess error:
    ERROR: Command errored out with exit status 1:
     command: /home/mghenis/anaconda3/envs/taxdata-dev/bin/python -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-7ndz7san/rpy2/setup.py'"'"'; __file__='"'"'/tmp/pip-install-7ndz7san/rpy2/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base /tmp/pip-pip-egg-info-oqfbuzdl
         cwd: /tmp/pip-install-7ndz7san/rpy2/
    Complete output (2 lines):
    cffi mode: CFFI_MODE.ANY
    Error: rpy2 in API mode cannot be built without R in the PATH or R_HOME defined. Correct this or force ABI mode-only by defining the environment variable RPY2_CFFI_MODE=ABI
    ----------------------------------------
ERROR: Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.
```
Can you use the `conda` installers for `rpy2` and `tzlocal` instead? That worked for me, after adding `r` to the list of channels (`rpy2` isn't on `conda-forge`).

MaxGhenis commented:
You might still run into problems in the build though, since the `rpy2` section imports the `censReg` R package, which seems only to be on CRAN (probably won't be installed in Travis). `censReg` is used for the tobit regressions for [imputing charitable deductions](https://github.com/PSLmodels/taxdata/pull/314/files?file-filters%5B%5D=.py#diff-5c433e930b1216b744733b6bc2d5179a) without producing negative valued imputations (the only use of `rpy2`). Alternatives could be using Python code for tobit regressions (e.g. [this](https://github.com/jamesdj/tobit) is based on `censReg` but isn't a full package), or just using OLS and truncating at zero post-hoc. Either way we'd probably want to explore alternative imputation approaches in the future.

andersonfrailey commented:
Thanks for the tips, @MaxGhenis! I think I'm going to remove those two packages for now. I ended up using the parameters John used for imputations rather than the ones from censreg. I also had a bunch of trouble installing rpy2 locally recently and figured we could stick with John's until that got figured out.

andersonfrailey commented:
I feel comfortable merging this. Happy to answer any lingering questions about it before doing so.

MaxGhenis commented on cps_data/pycps/helpers.py:
```diff
@@ -0,0 +1,108 @@
+    array_first = True
+
+
+FILINGPARAMS = FilingParams()
```
not a constant for uppercase

MaxGhenis commented on cps_data/pycps/helpers.py:
```diff
@@ -0,0 +1,108 @@
+        return pd.read_csv(path, usecols=usecols, index_col=index_col)
+
+    # Set global variables so they can be accested later
+    global MCAID, MCARE, VB, SNAP, SSI, SS, HOUSING, TANF, UI, WIC
```
these should be lowercase

MaxGhenis commented on cps_data/pycps/helpers.py:
```diff
@@ -0,0 +1,108 @@
+
+FILINGPARAMS = FilingParams()
+# this index is used to access the filing parameter values for the correct year
+PUF_YR_IDX = PUF_YEAR - FILINGPARAMS.label_grid["year"][0]
```
this and below should be lowercase

MaxGhenis commented on cps_data/pycps/impute.py:
```diff
@@ -0,0 +1,262 @@
+        data["mars"] == 2, 1, 0
+    )
+    # cap family size at 5 to match with the PUF
+    # data["fam_size"] = np.minimum(data["XTOT"], 5)
```
is family size being capped elsewhere? if not could you uncomment the code, and possibly leave a comment re: why it's not being done?

andersonfrailey replied:
Good catch! Should be uncommented.

MaxGhenis commented on cps_data/pycps/impute.py:
```diff
@@ -0,0 +1,262 @@
+    dpad_indicator = np.where(
+        (data["e00900"] > 0) | (data["rents"] > 0), 1, 0
+    )
+    dpad_bins = [
```
uppercase this and other constants in this file (dpad_probs, dpad_bases)

MaxGhenis commented on cps_data/pycps/impute.py:
```diff
@@ -0,0 +1,262 @@
+    ]
+
+    data["CHARITABLE"] = tobit(
+        data, ols_betas["char_ols"], TOBIT_XVARS, 48765.45, 1.
```
Where this 48765.45 from?

And in general there are a lot of hard-coded numbers in this file. Are they documented? Ideally I think these would be their own defined constants rather than entered in functions, or even better would be loading them from CSVs.

andersonfrailey replied:
All of the hard coded numbers come from John and can be found in [this](https://github.com/PSLmodels/taxdata/blob/master/cps_data/State%20Database%20-%202015/Production%20File/Imputations/Impute3TobitV4.sas) file. I carried them over to be consistent with what we've been doing for imputations up to this point. Agreed that it's less than ideal. I'd like for one of the next TaxData projects to be revisiting all of the imputations we do on the CPS to make them more transparent, reproducible, easier to update, etc. Getting rid of all the hard coded numbers would be part of that work.

MaxGhenis commented on cps_data/pycps/splitincome.py:
```diff
@@ -0,0 +1,61 @@
+    np.random.seed(79)
+
+    # Qualified dividends
+    all_qualified_prob = 0.429  # % of units where all dividends are qualified
```
uppercase (others too in this file)

MaxGhenis commented on cps_data/pycps/targeting.py:
```diff
@@ -0,0 +1,85 @@
+        "A00650": ["e00650"],
+        "A00900": ["e00900p", "e00900s"],
+        "A02300": ["e02300"],
+        # "A01000": ["CGAGIX"],
```
why is this commented?

andersonfrailey replied:
It incorrectly specified which target to use for `CGAGIX`. Should've been deleted

MaxGhenis commented on cps_data/pycps/validation.py:
```diff
@@ -0,0 +1,107 @@
+
+    # compare income variables
+    # add up income totals in the raw cps
+    # cps_totals = dict(
```
why commented?

andersonfrailey replied:
I was re-factoring the validation scripts and didn't want to delete it just in case things didn't work. Forgot to go back and delete it later. Gone now 👍🏾

MaxGhenis commented on cps_data/pycps/validation.py:
```diff
@@ -0,0 +1,107 @@
+        # values is greater than .1 because for some reason
+        # interest income would raise an error even though the totals
+        # appeared to be the same
+        # error = cps_sum < tc_sum and abs(cps_sum - tc_sum) > .1
```
why commented? also looks like the preceding comment refers to this, should that be deleted?

andersonfrailey replied:
Yep. This test is taken care of on the next line. Deleting

MaxGhenis commented on cps_stage2/stage2.py:
```diff
@@ -1,70 +1,84 @@
     # DataFrame for holding each year's weights
     weights = pd.DataFrame()
+    print("Solving for 2014")
     weights['WT2014'] = solve_lp_for_year(cps, stage_1_factors,
```
a loop here would be more concise

MaxGhenis commented on tests/test_data.py:
```diff
@@ -182,44 +193,43 @@ def check_cps_benefits(data):
               'tanf', 'housing', 'vet', 'other']
-    expect_minben = 0
     # specify expected benefit statistics in CPS data
     expect_ben_stat = dict()
```
uppercase for constant (ideally a CSV)

MaxGhenis commented on tests/test_data.py:
```diff
@@ -182,44 +193,43 @@ def check_cps_benefits(data):
     """
     bnames = ['mcare', 'mcaid', 'ssi', 'snap', 'wic',
```
uppercase constant

MaxGhenis commented on tests/test_weights.py:
```diff
@@ -50,7 +50,7 @@ def test_weights(kind, cps_weights, puf_weights,
                                         weights[col].max(), max_weight))
     # test sum of weights (in millions) for each year
     min_weight_sum = 149
```
uppercase (also max_weight_sum)

andersonfrailey commented:
Thanks for the review, @MaxGhenis! Pushing updates in response shortly

andersonfrailey commented:
Pushed everything new up

andersonfrailey commented:
Are there any more comments on this or can I go ahead and merge?

@MattHJensen @MaxGhenis @Peter-Metz @hdoupe

MattHJensen commented:
@andersonfrailey, this is very exciting!  

Would you mind (re) summarizing your work when you have an opportunity? (The top summary comment is somewhat out of date now thanks to all of your updates and Max's feedback!)

MattHJensen commented on cps_data/pycps/finalprep.py:
```diff
@@ -0,0 +1,285 @@
+    return var
+
+
+def adjust(data, targets):
```
@andersonfrailey, it doesn't appear that `adjust()` or `adjust_helper()` is taking into account the filer/non-filer distinction when creating the adjustment factors. 

`target()` seems very similar to `adjust,` just focussed on a wider variety of state-based adjustments. (@andersonfrailey, if I am missing anything in that assumption I'd love a correction). 

`target()` calculates the adjustment factor as the difference between the filer population's sum and the SOI target, and then it applies the adjustment factor across both filers and non-filers. This generally makes sense to me (and seems like a fine approach for this PR since it doesn't seem to be a departure from the SAS code), but I want to think more about whether it ends up skewing the filer/non-filer count one way or the other.

andersonfrailey replied:
@MattHJensen, in addition to focusing on a wider variety of state-based adjustments, `target()` actually changes the total value of the targeted variables while `adjust()` only changes the distribution of the targeted variable in the data. We don't make a distinction between filers and non-filers in `adjust()` because we aren't changing the totals.

Neither of these would change the unweighted total of filers/non-filers because we make that distinction before any adjustments are made, but they could affect weighted sums after the stage 2 adjustments.

MattHJensen replied:
Thanks Anderson. Resolving.

andersonfrailey commented:
>Would you mind (re) summarizing your work when you have an opportunity?

Happy to!

I'm working on much more detailed documentation of what's happening here, but the gist is that this PR ports the code used to create `cps.csv.gz` from SAS to Python. The general steps we take are:
1. Create tax units from the 2013, 2014, 2015 CPS March supplement files. The CPS data information on marital status, if/by whom an individual is claimed as a dependent, and filing status. We use that to form tax units within each household. The tax units from each of these individual files are combined into one dataset.
2. We split up certain income data we have in the CPS to match how it's classified in the IRS Public Use File. For example, dividend income reported in the CPS lumps qualified and ordinary dividends in one variable. We split what's reported in the CPS based on the ratio of the two in the 2011 PUF.
3. We target some income variables so that the weighted totals among those determined to be tax filers in the CPS match what the [IRS reports](https://www.irs.gov/statistics/soi-tax-stats-state-data-at-a-glance-fy-2014).
4. We impute a number of itemized deductions that are not found in the CPS to round out the data in the file.
4. As we're cleaning up the file, we target the distribution of interest, dividend, and business income to match those found in IRS data.

All of the tax units also have benefit program data imputed from C-TAM.

A couple of additional files are added by the PR unrelated to the logic behind creating tax units. `validation.py` is used to validate the income totals and person counts in each tax unit. `transform_sas.py` and `template.txt` are used to translate the SAS code provided by NBER to convert the CPS files from a .dat format to something useable.

MattHJensen commented:
Thanks a lot for the overview Anderson. I am ever so happy to be leaving SAS behind and entering a new realm of continuous improvement in Python thanks to this work!

andersonfrailey commented:
>I am ever so happy to be leaving SAS behind and entering a new realm of continuous improvement in Python

You and me both! I'm going to tag the current version of taxdata just in case we need to revisit the SAS code for whatever reason, and then merge this PR at the end of the day if there are no objections.
